infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/project/OpenShiftDeploymentConfigs.java

```diff
@@ -0,0 +1,223 @@
+      String nameSuffix = pod.getMetadata().getLabels().get("che.original_name");
+      deploymentResource =
+          clientFactory
+              .create(workspaceId)
+              .extensions()
+              .deployments()
+              .inNamespace(namespace)
+              .withName(String.format("%s.%s", deploymentName, nameSuffix));
+    }
+    if (deploymentResource.get() == null) {
+      throw new InfrastructureException(
+          String.format("Could not get DeploymentConfig or Pod with name %s", name));
+    }
+    deploymentResource.cascading(true).delete();
+    CompletableFuture<Void> future = new CompletableFuture<Void>();
+    future.complete(null);
+    return future;
+  }
+
+  private Pod awaitAndGetPod(Deployment dc) throws InfrastructureException {
+    String deploymentName = dc.getMetadata().getName();
+    try {
+      clientFactory
+          .create(workspaceId)
+          .extensions()
+          .deployments()
+          .withName(deploymentName)
+          .waitUntilReady(DEPLOYMENT_CONFIG_READY_TIMEOUT_MIN, TimeUnit.MINUTES);
+      return get(deploymentName)
+          .orElseThrow(
+              () ->
+                  new InfrastructureException(
+                      String.format("Failed to create DeploymentConfig %s", deploymentName)));
+    } catch (InterruptedException e) {
+      throw new InfrastructureException(
+          String.format("Timeout while creating DeploymentConfig %s", deploymentName));
+    }
+  }
+
+  /**
+   * Returns the name of a specified Pod given either the actual Pod name or the name of the
+   * DeploymentConfig that controls it. <br>
+   * This is necessary because we are trying to transparently wrap Pods in DeploymentConfigs;
+   * attempting to create a Pod named {@code testPod} will result in a DeploymentConfig with name
+   * {@code testPod}, which will in turn create a Pod named e.g {@code testPod-1-xxxxx}.
+   *
+   * @param name Pod or DeploymentConfig name
+   * @return the name of the intended pod.
+   * @see
+   */
+  private String getPodName(String name) throws InfrastructureException {
+    if (clientFactory.create(workspaceId).pods().inNamespace(namespace).withName(name).get()
+        != null) {
+      return name;
+    }
+    Deployment deployment =
+        clientFactory
+            .create(workspaceId)
+            .extensions()
+            .deployments()
+            .inNamespace(namespace)
+            .withName(name)
+            .get();
+    Map<String, String> selector = deployment.getSpec().getSelector().getMatchLabels();
+    List<Pod> pods =
+        clientFactory
+            .create(workspaceId)
+            .pods()
+            .inNamespace(namespace)
+            .withLabels(selector)
+            .list()
+            .getItems();
+    if (pods.isEmpty()) {
+      throw new InfrastructureException(String.format("Failed to find pod with name %s", name));
+    } else if (pods.size() > 1) {
+      throw new InfrastructureException(
```

Comment:
Could you go into this scenario a little more? As I see it, there shouldn't be a situation in which multiple pods are within the same deployment, and so if that does occur it suggests an error state somewhere along the way. 

I'm trying to work something out where we check for the rolling state, but it seems like it would add quite a bit of logic, so for clarity's sake I want to understand the motivation.